jetstream/config.py

```diff
@@ -62,64 +73,91 @@
 @attr.s(auto_attribs=True)
 class Platform:
     config_spec_path: PathLike
-    metrics_module: Optional[ModuleType]
-    segments_module: Optional[ModuleType]
     enrollments_query_type: str
     # app_id to use to validate Outcomes.
     validation_app_id: str
+    metrics_module: Optional[ModuleType] = attr.ib(
+        default=None
+    )  # kik-kik: moved to the end since this is "optional"
+    segments_module: Optional[ModuleType] = attr.ib(
+        default=None
+    )  # kik-kik: moved to the end since this is "optional"
 
 
-PLATFORM_CONFIGS = {
-    "firefox_desktop": Platform(
-        Path(__file__).parent / "config" / "default_metrics.toml",
-        mozanalysis.metrics.desktop,
-        mozanalysis.segments.desktop,
-        "normandy",
-        "firefox-desktop",
-    ),
-    "fenix": Platform(
-        Path(__file__).parent / "config" / "fenix.toml",
-        mozanalysis.metrics.fenix,
-        None,
-        "glean-event",
-        "org.mozilla.fenix",
-    ),
-    "firefox_ios": Platform(
-        Path(__file__).parent / "config" / "firefox_ios.toml",
-        mozanalysis.metrics.firefox_ios,
-        None,
-        "glean-event",
-        "org.mozilla.ios.FirefoxBeta",
-    ),
-    "focus_android": Platform(
-        Path(__file__).parent / "config" / "focus_android.toml",
-        mozanalysis.metrics.focus_android,
-        None,
-        "glean-event",
-        "org.mozilla.focus",
-    ),
-    "klar_android": Platform(
-        Path(__file__).parent / "config" / "klar_android.toml",
-        mozanalysis.metrics.klar_android,
-        None,
-        "glean-event",
-        "org.mozilla.klar",
-    ),
-    "focus_ios": Platform(
-        Path(__file__).parent / "config" / "focus_ios.toml",
-        mozanalysis.metrics.focus_ios,
-        None,
-        "glean-event",
-        "org.mozilla.ios.Focus",
-    ),
-    "klar_ios": Platform(
-        Path(__file__).parent / "config" / "klar_ios.toml",
-        mozanalysis.metrics.klar_ios,
-        None,
+PARENT_DIR = Path(__file__).parent
+CONFIG_DIRECTORY = PARENT_DIR / "config"
+
+
+def _generate_platform_config(config: MutableMapping[str, Any]) -> Dict[str, Platform]:
+    """
+    Takes platform configuration and generated platform object map
+    """
+
+    _module_mapping_metrics = {
+        "mozanalysis.metrics.desktop": mozanalysis.metrics.desktop,
+        "mozanalysis.metrics.fenix": mozanalysis.metrics.fenix,
+        "mozanalysis.metrics.focus_android": mozanalysis.metrics.focus_android,
+        "mozanalysis.metrics.klar_android": mozanalysis.metrics.klar_android,
+        "mozanalysis.metrics.firefox_ios": mozanalysis.metrics.firefox_ios,
+        "mozanalysis.metrics.focus_ios": mozanalysis.metrics.focus_ios,
+        "mozanalysis.metrics.klar_ios": mozanalysis.metrics.klar_ios,
+        "None": None,
+        "none": None,
+        None: None,
```

Comment:
This gives me an idea:

So ideally when adding a new platform we want to make as few (ideally none) code changes as possible. Having the mapping logic here will always require making updates.

In the current configuration logic there is an existing [`_lookup_name`](https://github.com/mozilla/jetstream/blob/fc3aa19e93a8b3a286fc1af169ca6a22a55ec599/jetstream/config.py#L150-L165) method which can transform a string into a class instance that exists in mozanalysis. We use it for resolving `DataSources`, `Metrics` and `Segments` that have been defined in mozanalysis and users reference in their configurations.

It would be cool to do something similar here, to use the configuration strings and resolve the corresponding modules. And I _think_ it might be possible. What I imagine is we'll need to import all available `mozanalysis.metrics` modules like:

```
from mozanalysis.metrics import *    # noqa: F403
```

(yep, flake8 will complain here, because usually it's not best practice, but the gain in convenience would be worth it)

To do the `import *`, we'll need to [define `__all__`](https://docs.python.org/3/tutorial/modules.html#importing-from-a-package) in the `mozanalysis.metrics` module source code: https://github.com/mozilla/mozanalysis/blob/main/src/mozanalysis/metrics/__init__.py. Ideally, available packages will be determined dynamically for `__all__`

Then we should be able to do something similar to `_lookup_name` when resolving the configuration:

```
module = getattr(mozanalysis.metrics, name_from_config)
```

Similarly, if this actually works we should do it also for `Segments`.

We could even get rid of having to explicitly specify `metrics_module` and `segments_module` altogether and derive it from the platform name (we'll need some mapping for `desktop` <-> `firefox_desktop`).

wdyt?

Reply:
ok so I initially played around with `__all__` but due to naming clashes I couldn't figure out how to resolve it. I came up with a slightly different solution of dynamically detecting segments and metrics modules available in the package and loading them in. Let me know what you think about that approach.